src/Hedgehog/Internal/Report.hs

```diff
@@ -0,0 +1,611 @@
+ppReproduce name size seed =
+  WL.vsep [
+      markup ReproduceHeader
+        "This failure can be reproduced by running:"
+    , gutter ReproduceGutter . markup ReproduceSource $
+        "recheck" <+>
+        WL.text (showsPrec 11 size "") <+>
+        WL.text (showsPrec 11 seed "") <+>
+        maybe "<property>" WL.text name
+    ]
+
+mergeLine :: Semigroup a => Line a -> Line a -> Line a
+mergeLine (Line x no src) (Line y _ _) =
+  Line (x <> y) no src
+
+mergeDeclaration :: Semigroup a => Declaration a -> Declaration a -> Declaration a
+mergeDeclaration (Declaration file line name src0) (Declaration _ _ _ src1) =
+  Declaration file line name $
+  Map.unionWith mergeLine src0 src1
+
+mergeDeclarations :: Semigroup a => [Declaration a] -> [Declaration a]
+mergeDeclarations =
+  Map.elems .
+  Map.fromListWith mergeDeclaration .
+  fmap (\d -> ((declarationFile d, declarationLine d), d))
+
+ppFailure :: MonadIO m => Maybe String -> Failure -> m (Doc Markup)
+ppFailure name (Failure size seed _ inputs0 location0 msgs) = do
+  mlocation <- ppFailureLocation location0
+  minputs <- traverse ppFailedInput inputs0
+
+  let
+    decls =
+      mergeDeclarations .
+      catMaybes $
+        mlocation : minputs
+
+  pure . WL.indent 2 . WL.vsep . WL.punctuate WL.line $ [
+      WL.vsep . WL.punctuate WL.line $ fmap ppDeclaration decls
+    ] <> (if null msgs then [] else [WL.vsep $ fmap WL.text msgs]) <> [
+      ppReproduce name size seed
+    ]
+
+ppName :: Maybe String -> Doc a
+ppName = \case
+  Nothing ->
+    "<interactive>"
+  Just name ->
+    WL.text name
+
+ppReport :: MonadIO m => Maybe String -> Report -> m (Doc Markup)
+ppReport name (Report successes discards status) =
+  case status of
+    Failed failure -> do
+      pfailure <- ppFailure name failure
+      pure . WL.vsep $ [
+          icon FailedIcon '✗' . WL.annotate FailedHeader $
+            ppName name <+>
+            "failed after" <+>
+            ppSuccessCount successes <>
+            ppShrinkDiscard (failureShrinks failure) discards <>
+            "."
+        , mempty
+        , pfailure
+        , mempty
+        ]
+    GaveUp ->
+      pure . icon GaveUpIcon '⚐' . WL.annotate GaveUpHeader $
+        ppName name <+>
+        "gave up after" <+>
+        ppDiscardCount discards <>
+        ", passed" <+>
+        ppSuccessCount successes <>
+        "."
+    OK ->
+      pure . icon SuccessIcon '✓' . WL.annotate SuccessHeader $
+        ppName name <+>
+        "passed" <+>
+        ppSuccessCount successes <>
+        "."
+
+useColor :: MonadIO m => m Bool
+useColor = do
+  -- FIXME This is pretty savage, not sure how else to provide this while
+  -- FIXME maintain a seamless experience. Maybe we should just go for a
+  -- FIXME HEDGEHOG_COLOR environment variable instead.
+  args <- liftIO getArgs
+  if elem "--color" args then
+    pure True
+  else if elem "--no-color" args then
+    pure False
+  else
+    liftIO $ hSupportsANSI stdout
```

Comment:
Agree with your env variable suggestion, this `getArgs` is pretty brutal.

`hSupportANSI` will be correct for 99% of users though 👍